### PR TITLE
[REVIEW] Fix error in binops when only one input col has null mask

### PR DESCRIPTION
**Pull request by devavret**
Fix for #696 . 
New approach is to simply copy the mask from the one input column that has it.
In case neither input columns have a mask, the output mask is set to all 1.

### CODE REVIEW

**Comment by kkraus14** [comment]
@devavret looks like there's no unit tests for this

File: cpp/src/binary/binary_ops.cu [hunk 1]
```diff
@@ -510,17 +510,33 @@ DEF_BITWISE_IMPL_GROUP(xor, DeviceBitwiseXor)
...
+    // TODO: this will need to be changed when gdf_valid_type is changed to 4 byte
+    if ( x.data == nullptr && y.data != nullptr ) {
+        CUDA_TRY( cudaMemcpy(z.data, y.data, y.size, cudaMemcpyDeviceToDevice) );
+        return GDF_SUCCESS;
+    } 
+    else if ( x.data != nullptr && y.data == nullptr ) {
+        CUDA_TRY( cudaMemcpy(z.data, x.data, x.size, cudaMemcpyDeviceToDevice) );
+        return GDF_SUCCESS;
+    } 
+    else if ( x.data == nullptr && y.data == nullptr ) {
```
Comment (jrhemstad) [inline cpp/src/binary/binary_ops.cu]: If both inputs don't have a bitmask, there's no reason to allocate one for the output. The bitmask is optional after all.

File: cpp/src/binary/binary_ops.cu [hunk 2]
```diff
@@ -510,17 +510,33 @@ DEF_BITWISE_IMPL_GROUP(xor, DeviceBitwiseXor)
...
 gdf_error gdf_validity_and(gdf_column *lhs, gdf_column *rhs, gdf_column *output) {
-    GDF_REQUIRE ( lhs->valid && rhs->valid && output->valid, GDF_VALIDITY_MISSING);
+    GDF_REQUIRE ( output->valid, GDF_VALIDITY_MISSING);
     gdf_column x = gdf_validity_column(*lhs);
     gdf_column y = gdf_validity_column(*rhs);
     gdf_column z = gdf_validity_column(*output);
+    
+    // TODO: this will need to be changed when gdf_valid_type is changed to 4 byte
+    if ( x.data == nullptr && y.data != nullptr ) {
+        CUDA_TRY( cudaMemcpy(z.data, y.data, y.size, cudaMemcpyDeviceToDevice) );
```
Comment (jrhemstad) [inline cpp/src/binary/binary_ops.cu]: You'll also need to update the output column's null count.

File: cpp/src/binary/binary_ops.cu [hunk 3]
```diff
@@ -510,17 +510,33 @@ DEF_BITWISE_IMPL_GROUP(xor, DeviceBitwiseXor)
...
+        return GDF_SUCCESS;
+    } 
+    else if ( x.data != nullptr && y.data == nullptr ) {
+        CUDA_TRY( cudaMemcpy(z.data, x.data, x.size, cudaMemcpyDeviceToDevice) );
```
Comment (jrhemstad) [inline cpp/src/binary/binary_ops.cu]: Update output's null count

File: cpp/src/binary/binary_ops.cu [hunk 4]
```diff
@@ -510,17 +510,33 @@ DEF_BITWISE_IMPL_GROUP(xor, DeviceBitwiseXor)
...
+        return GDF_SUCCESS;
+    } 
+    else if ( x.data == nullptr && y.data == nullptr ) {
+        CUDA_TRY( cudaMemset(z.data, 0xff, x.size) );
+        return GDF_SUCCESS;
+    }
+    
     gdf_bitwise_and_i8(&x, &y, &z);
```
Comment (jrhemstad) [inline cpp/src/binary/binary_ops.cu]: Need to update output's null count after the AND.

File: cpp/src/binary/binary_ops.cu (hunk 1 above)
Comment (devavret) [reply cpp/src/binary/binary_ops.cu]: cuDF currently only calls this if it has allocated a null mask for out. Meaning they want it filled.

Comment (jrhemstad) [reply cpp/src/binary/binary_ops.cu]: Understood, then this is fine for now. 

Ultimately, this means that there's a problem in how cuDF wraps the binary operations, but we've already discussed this and it's outside the scope of this PR.

File: cpp/src/binary/binary_ops.cu (hunk 2 above)
Comment (devavret) [reply cpp/src/binary/binary_ops.cu]: That's not what the name of the function says. I'll add another function that does the count update in a separate PR where I'm fixing the fact that it takes two api calls to do a binary op and null mask calculation.

File: cpp/src/binary/binary_ops.cu (hunk 3 above)
Comment (devavret) [reply cpp/src/binary/binary_ops.cu]: cuDF also updates the null count right after calling this, using a number kernel.

**Comment by devavret** [comment]
Added unit tests but they're pytests, not gtests.

**Comment by kkraus14** [comment]
> Added unit tests but they're pytests, not gtests.

Works for me but I can't speak for @jrhemstad 😄

File: cpp/src/binary/binary_ops.cu (hunk 2 above)
Comment (jrhemstad) [reply cpp/src/binary/binary_ops.cu]: Sounds good, can be addressed in another PR.

**Comment by jrhemstad** [comment]
> > Added unit tests but they're pytests, not gtests.
> 
> Works for me but I can't speak for @jrhemstad

This function is an aberration that needs to be gutted anyways, so I don't care about testing it right now. 

@devavret and I have already discussed how to make things better, but it will be in a different PR.

**Pull request merged**